HXCPP_GC_GENERATIONAL fixes and improvements

sibest opened the pull request:
As requested, i've created a pull request with my changes. 

I've noticed that 0 bytes allocations were going into the nursery, I have mainly fixed this issue that was causing random crashes. Plus, i've implemented the possibility to decrease the line length (i've set to 6 bits per default). I've disabled the large recycle pool and created a pool of 5 large list to mainly collect only the less used ones.

**If you are using lime, you have to rebuild it to match the new inline functions**

Aurel300 commented on include/hx/CFFINekoLoader.h:
```diff
@@ -230,25 +230,26 @@ int api_val_strlen(neko_value  arg1)
 }
 void api_buffer_set_size(neko_buffer inBuffer,int inLen) { 
-   NEKO_NOT_IMPLEMENTED("api_buffer_set_size");
+  // NEKO_NOT_IMPLEMENTED("api_buffer_set_size");
```
What is this? I suppose these Neko changes are not actually related to the GC improvements?

sibest replied:
i was building my project with "openfl build windows" and it was giving the error api_buffer_set_size NOT IMPLEMENTED while trying to copy the ico files into the executable. I've reverted some changes in the CFFINekoLoader to let it work.

Aurel300 replied:
Still I don't think it should be part of this PR, it really sounds like an unrelated issue.

sibest replied:
Indeed it's unrelated

Aurel300 commented on include/hx/GC.h:
```diff
@@ -277,24 +279,19 @@ void GCPrepareMultiThreaded();
 
 // Inline code tied to the immix implementation
-
+ 
```
Please try to reduce the whitespace changes.

```suggestion

```

Aurel300 commented on include/hx/Object.h:
```diff
@@ -166,11 +166,11 @@ class HXCPP_EXTERN_CLASS_ATTRIBUTES Object
             BadImmixAlloc();
          #endif
-
+         
```
```suggestion

```

Aurel300 commented on include/hx/Object.h:
```diff
@@ -166,11 +166,11 @@ class HXCPP_EXTERN_CLASS_ATTRIBUTES Object
 
       #else // Not HX_USE_INLINE_IMMIX_OPERATOR_NEW ...
-
+            
```
This file isn't actually changed …

```suggestion

```

sibest replied:
Yes sorry, i made cleanup for Immix.cpp but didn't check the include files (didn't sleep this night = P)

Aurel300 commented on src/hx/gc/Immix.cpp:
```diff
@@ -451,21 +454,22 @@ extern void scriptMarkStack(hx::MarkContext *);
    typedef unsigned short BlockIdType;
 #endif
 
+#define IMMIX_START_MASK				(IMMIX_LINE_LEN-1)
```
```suggestion
#define IMMIX_START_MASK        (IMMIX_LINE_LEN-1)
```

Aurel300 commented on src/hx/gc/Immix.cpp:
```diff
@@ -451,21 +454,22 @@ extern void scriptMarkStack(hx::MarkContext *);
 #define IMMIX_LINES             (1<<IMMIX_LINE_COUNT_BITS)
 
-
+#define IMMIX_BLOCKID_SIZE			sizeof(BlockIdType)
```
```suggestion
#define IMMIX_BLOCKID_SIZE      sizeof(BlockIdType)
```

Aurel300 commented on include/hx/GC.h:
```diff
@@ -103,8 +109,6 @@ hx::Object *__hxcpp_id_obj(int);
 
-
-
 namespace hx
```
```suggestion


namespace hx
```

Aurel300 commented on include/hx/GC.h:
```diff
@@ -277,24 +279,18 @@ void GCPrepareMultiThreaded();
-
-// Each line ast 128 bytes (2^7)
-#define IMMIX_LINE_BITS    7
+#define IMMIX_LINE_BITS    6
```
```suggestion

// Each line spans 64 bytes (2^6)
#define IMMIX_LINE_BITS    6
```

Aurel300 commented on include/hx/GC.h:
```diff
@@ -331,12 +327,11 @@ EXTERN_FAST_TLS_DATA(StackContext, tlsStackContext);
 extern void BadImmixAlloc();
 
-
 class ImmixAllocator
```
```suggestion

class ImmixAllocator
```

Aurel300 commented on include/hx/GC.h:
```diff
@@ -277,24 +279,18 @@ void GCPrepareMultiThreaded();
 
 // Inline code tied to the immix implementation
-
 namespace hx
```
```suggestion

namespace hx
```

Aurel300 commented on include/hx/GC.h:
```diff
@@ -277,24 +279,18 @@ void GCPrepareMultiThreaded();
 
 #define HX_USE_INLINE_IMMIX_OPERATOR_NEW
-
 //#define HX_STACK_CTX ::hx::ImmixAllocator *_hx_stack_ctx =  hx::gMultiThreadMode ? hx::tlsImmixAllocator : hx::gMainThreadAlloc;
```
```suggestion

//#define HX_STACK_CTX ::hx::ImmixAllocator *_hx_stack_ctx =  hx::gMultiThreadMode ? hx::tlsImmixAllocator : hx::gMainThreadAlloc;
```

Aurel300 commented on include/hx/GC.h:
```diff
@@ -359,7 +356,29 @@ class ImmixAllocator
    // These allocate the function using the garbage-colleced malloc
    inline static void *alloc(ImmixAllocator *alloc, size_t inSize, bool inContainer, const char *inName )
    {
+
+   	  if (!inContainer) {
+   	  	
+   	  	// a non-object with 0 bytes length ? it would make the alloc memory type
+   	  	// check mask in mistake
+   	  	if (inSize == 0)
+   	  		 return 0;
+   	  		
+   	  	inSize = (inSize+3)&~3;
+   	  }
+   	  else
+   	  {
+	      #if defined(HXCPP_VISIT_ALLOCS) && defined(HXCPP_M64)
+	      // Make sure we can fit a relocation pointer
+	      inSize += 4;
+	      #endif
+   	  }
+   	  	
+   	 // if (inContainer == false && inSize > IMMIX_LARGE_OBJ_SIZE)
+   	  	
+   	    	 
       #ifdef HXCPP_GC_NURSERY
+      
```
Please try to be consistent with the indentation here, I believe hxcpp uses 3 spaces for a tab. (Don't ask me why…)

sibest replied:
You are right, I'm using UEDIT32 and I think that hughsando is using MSVC.

sibest replied:
Plus // if (inContainer == false && inSize > IMMIX_LARGE_OBJ_SIZE) this should be removed

sibest commented:
Currently with this pull 0-byte allocations are not allowed for non-containers. The class Socket and File allocs 0-length strings sometimes. I've changed the code and added the flag
#define IMMIX_ALLOC_IS_EMPTY       0x00400000

to allow 0-bytes non-containers in the nursery.  I will upload it later.

sibest commented:
Discard this pull, I’m going to make a new one
